### ISSUE TITLE
PATCH Users in the ADMIN REST API missing

### ISSUE THREAD

**Issue by TeixeiraNelson** [keycloak/keycloak#19691]
### Description

I am using Keycloak docker image 21.0.2 (the latest one).

I make a PUT request to this URL :  http://localhost:8080/admin/realms/custom-realm/users/14905db1-5229-4110-bfbb-df13845a4f52 and with this JSON body : { "attributes": { "organization_id": "1" } } to edit a user's attributes.

The problem is that this wipes out all user attributes and i want it to keep the old ones as they are used for other functionalities and cannot be wiped out.

I just want to append or replace 'organization_id' value in the list of user attributes.

I tried to update user attributes through the ADMIN REST API but I think a PATCH method is missing...

### Motivation

We should be abled to update specific fields if we want/need to through the REST API

### Details

Add a PATCH method to the REST API

**Comment by mrjwt** [comment]
Hi. What is the status of this issue? When are you planning to add this?

**Comment by SeWieland** [comment]
I think the maintainers are aware of those issues, but the priority seems to be quite low: https://github.com/keycloak/keycloak/discussions/8983

**Comment by jonkoops** [comment]
Yes, we would love to create a new Admin API eventually. But currently there are higher priorities we need to focus on.

**Comment by lmller** [comment]
It seems like at least the java admin client actually perform a PATCH when `update(UserRepresentation)` is called. At least it's not changing the whole user if I only change the `enabled` flag.
This is nice, as it's the behavior I want, but it's not really documented anywhere.... Can someone confirm this?

**Comment by Lordfirespeed** [comment]
Would the Keycloak maintainer team be open to contributions to this issue? 

I would have thought updating/setting a user attribute from a client application is a very common use-case. As things stand, one must be careful to provide a list containing all existing user attributes, or they will be removed ... 

An additional route like `PATCH /admin/realms/{realm}/users/{id}/attributes` would also fit the bill.

**Comment by dalbani** [comment]
Keycloak doesn't keep track of the last modified date of a user or its attributes AFAIK.
So I think `If-Unmodified-Since` is not really an option, is it?
But I guess `If-Match` would be a good fit as it would be possible for Keycloak to calculate an ETag of a user's attributes.

**Comment by altanbgn** [comment]
Any update on this?

**Comment by Lordfirespeed** [comment]
> Any update on this?

I would still like to contribute, but don't see the point in doing so without confirmation that my contributions would be considered

**Comment by disbeliefff** [comment]
Any update on this?

**Comment by dontbug** [comment]
We're also very interested in adding PATCH support to the Keycloak Admin REST API for partial user updates—this would prevent accidental overwriting of existing attributes and enable safer, more efficient integrations in multi-system environments where concurrent updates are common.

As a temporary workaround, we're emulating PATCH behavior by using Redis to implement atomic locking at the per-user level. While the lock is held, we fetch the current user data via GET, merge in only the desired changes (e.g., updating a single attribute like 'organization_id'), and then submit the full updated object via PUT. This ensures no data loss from other attributes during the operation.

Implementing native PATCH would greatly simplify this and reduce reliance on external tools like Redis for concurrency control, making Keycloak even more robust for production use.

**Comment by ramin-n** [comment]
> Yes, we would love to create a new Admin API eventually. But currently there are higher priorities we need to focus on.

I think it's fair to expect that, after a couple of years, such a basic requirement would have been prioritized in development. Would you agree? Even a small patch endpoint to set user attributes would go a long way.

**Comment by johnnagro** [comment]
I'd like to work on this. I've written up a concrete proposal (an opt-in `PATCH` with `application/merge-patch+json`, leaving PUT unchanged) with a few open questions for maintainers here: **[discussion link](https://github.com/keycloak/keycloak/discussions/53172)**
